src/zenml/stack/stack_component.py

```diff
@@ -63,6 +67,156 @@ class StackComponent(BaseModel, ABC):
+                    f"Passing the stack component attribute `{key}` as a "
+                    "secret reference is not allowed as additional validation "
+                    "is required for this attribute."
+                )
+
+        super().__init__(**kwargs)
+
+    def __custom_getattribute__(self, key: str) -> Any:
+        """Returns the (potentially resolve) attribute value for the given key.
+
+        An attribute value may be either specified directly, or as a secret
+        reference. In case of a secret reference, this method resolves the
+        reference and returns the secret value instead.
+
+        Args:
+            key: The key for which to get the attribute value.
+
+        Raises:
+            RuntimeError: If the stack component is not part of the active
+                stack, or the active stack is missing a secrets manager.
+            KeyError: If the secret or secret key don't exist.
+
+        Returns:
+            The (potentially resolved) attribute value.
+        """
+        value = super().__getattribute__(key)
+
+        if secret_utils.is_secret_reference(value):
+            from zenml.repository import Repository
+
+            stack = Repository().active_stack
+
+            # A stack component can be part of many stacks, and currently a
+            # secrets manager is associated with a stack. This means we're
+            # not able to identify the 'correct' secrets manager that the user
+            # wanted to resolve the secrets in a general way. We therefore
+            # limit secret resolving to components of the active stack.
+            if stack.components[self.TYPE] != self:
+                raise RuntimeError(
+                    f"Failed to resolve secret reference for attribute {key} "
+                    f"of stack component `{self}`: The stack component is not "
+                    "part of the active stack and therefore can't have it's "
+                    "secret references resolved."
+                )
+
+            secrets_manager = Repository().active_stack.secrets_manager
+            if not secrets_manager:
+                raise RuntimeError(
+                    f"Failed to resolve secret reference for attribute {key} "
+                    f"of stack component `{self}`: The active stack does not "
+                    "have a secrets manager."
+                )
+
+            secret_ref = secret_utils.parse_secret_reference(value)
+            try:
+                secret = secrets_manager.get_secret(secret_ref.name)
+            except KeyError:
+                raise KeyError(
+                    f"Failed to resolve secret reference for attribute {key} "
+                    f"of stack component `{self}`: The secret "
+                    f"{secret_ref.name} does not exist."
+                )
+
+            try:
+                secret_value = secret.content[secret_ref.key]
+            except KeyError:
+                raise KeyError(
```

Comment:
Right now the `SecretValidationLevel` is used to control **what** we should validate:
- Nothing
- Secret existence
- Secret and key existence

The severity is controlled by when the validation is running:
- When registering/updating a stack: At this point the stack might not even contain a secrets manager yet or the user might simply not have registered the stack. We only warn them that the secrets are missing.
- When running a pipeline: At this point the secrets must exist, otherwise the pipeline will fail.

I separated these two for increased flexibility, for example: Someone might only have `list` access on secrets, but not `read` access. In this case they want to set the `SecretValidationLevel` to `SECRET_EXISTS` (this doesn't read the secret to check whether the key exists), but they probably still want ZenML to raise an exception if the secret is missing before running a pipeline.

Not sure if this convinced you, I'm also happy to unify these two concepts but unsure what combinations we should support.

Reply:
I understand. I imagine this might create some confusion among users if not thoroughly documented, but we'll just have to  release it into the wild to see how it impacts users first.

Reply:
That's a good point, the documentation around that is completely missing, I'll add that as well.